base/common/src/main/java/com/netscape/certsrv/profile/ProfileData.java

```diff
@@ -273,18 +286,339 @@ public boolean equals(Object obj) {
+                                        descriptorElement.appendChild(syntaxElement);
+                                    }
+                                    if (descriptor.getConstraint() != null) {
+                                        Element mConstraintElement = document.createElement("mConstraint");
+                                        mConstraintElement.appendChild(document.createTextNode(descriptor.getConstraint()));
+                                        descriptorElement.appendChild(mConstraintElement);
+                                    }
+                                    if (descriptor.getDescription(Locale.getDefault()) != null) {
+                                        Element descriptionElement = document.createElement("mDescription");
+                                        descriptionElement.appendChild(document.createTextNode(descriptor.getDescription(Locale.getDefault())));
+                                        descriptorElement.appendChild(descriptionElement);
+                                    }
+                                    if (descriptor.getDefaultValue() != null) {
+                                        Element defaultValueElement = document.createElement("mDef");
+                                        defaultValueElement.appendChild(document.createTextNode(descriptor.getDefaultValue()));
+                                        descriptorElement.appendChild(defaultValueElement);
+                                    }
+                                    constraintElement.appendChild(descriptorElement);
+                                }
+                                if (pcv.getValue() != null) {
+                                    Element pcvValueElement = document.createElement("value");
+                                    pcvValueElement.appendChild(document.createTextNode(pcv.getValue()));
+                                    constraintElement.appendChild(pcvValueElement);
+                                }
+                                pcElement.appendChild(constraintElement);
+                            }
+                        }
+                    valueElement.appendChild(pcElement);
+                    }
+                }
+                policysetElement.appendChild(valueElement);
+                policysetsElement.appendChild(policysetElement);
+            }
+            pdElement.appendChild(policysetsElement);
+        }
+        if (link != null) {
+            Element linkElement = link.toDOM(document);
+            pdElement.appendChild(linkElement);
+        }
+        return pdElement;
+    }
+
+    public static ProfileData fromDOM(Element profileDataElement) throws DOMException, Exception {
+
+        ProfileData profileData = new ProfileData();
+        String id = profileDataElement.getAttribute("id");
+        profileData.setId(id);
+
+        NodeList classIdList = profileDataElement.getElementsByTagName("classId");
```

Comment:
Just a bit of caution, apparently `getElementsByTagName()` works recursively. Here the code will use the first element that it finds (i.e. `item(0)`), but if that element does not exist immediately under `<Profile>`, it might incorrectly pick up another element with the same name somewhere else in the tree.

Reply:
Yeah I had some issues with that. This is why sometimes I define a sub-element from the main `Element`, so that the `id` values lower down the hierarchy don't get mixed up with any "outer" `id`.